google/cloud/pubsub/publisher_connection_test.cc

```diff
@@ -342,34 +347,71 @@ TEST(PublisherConnectionTest, HandleTransientEnabledRetry) {
 using ::google::cloud::testing_util::DisableTracing;
 using ::google::cloud::testing_util::EnableTracing;
 using ::google::cloud::testing_util::InstallSpanCatcher;
+using ::google::cloud::testing_util::OTelContextCaptured;
 using ::google::cloud::testing_util::SpanNamed;
+using ::google::cloud::testing_util::ThereIsAnActiveSpan;
 using ::testing::IsEmpty;
 using ::testing::UnorderedElementsAre;
 
 TEST(MakePublisherConnectionTest, TracingEnabled) {
   auto span_catcher = InstallSpanCatcher();
   auto mock = std::make_shared<pubsub_testing::MockPublisherStub>();
   Topic const topic("test-project", "test-topic");
+
+  auto mock_message_batch =
+      std::make_shared<pubsub_testing::MockMessageBatch>();
+  std::mutex mu;
+  std::condition_variable cv;
+  bool is_publish_complete = false;  // ABSL_GUARDED_BY(mu)
+  EXPECT_CALL(*mock_message_batch, SaveMessage(_)).Times(1);
+  EXPECT_CALL(*mock_message_batch, Flush).WillOnce([&] {
+    // Batch sink span should still be active when Flush is called.
+    EXPECT_TRUE(ThereIsAnActiveSpan());
+    EXPECT_TRUE(OTelContextCaptured());
+    return [&](auto) {
+      // Batch sink span should no longer be active, but there will be another
+      // active span so instead we can test the otel context is no longer
+      // captured.
+      EXPECT_FALSE(OTelContextCaptured());
```

Comment:
Argh, I think I told you to add this, but I am not sure it should work. That is, I think it will break when we start pushing the context here: https://github.com/googleapis/google-cloud-cpp/blob/7cb1ec3ee393fefe431bea87a872063c78f722cc/google/cloud/pubsub/internal/publisher_tracing_connection.cc#L91

Any parent span of `BatchSink::AsyncPublish` produced by this test should still be in the OTelContext stack.

---

I think my matcher is too dumb. (it only checks if the stack is empty or not). Maybe we should do something like:
```cc
ASSERT_TRUE(OTelContextCaptured());
auto context = internal::CurrentOTelContext().back();
return [&](auto) {
  EXPECT_FALSE(internal::CurrentOTelContext(), Contains(context));
  // blah blah blah...
};
```

Reply:
I added it because I wanted the following to be true

```
  EXPECT_CALL(*mock_message_batch, Flush).WillOnce([&] {
    // Batch sink span should still be active when Flush is called.
    EXPECT_TRUE(ThereIsAnActiveSpan());
    return [&](auto) {
      // Batch sink span should no longer be active.
      EXPECT_FALSE(ThereIsAnActiveSpan());
      //  Wait until the lambda returned from Flush is executed to GetSpans.
      is_publish_complete.set_value();
    };
  });
```  
But the parent span was still active.

I think for the e2e test the matcher might not be a good fit. Removed.